samples/StaticFilesAuth/Startup.cs

```diff
@@ -114,81 +138,61 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, IAuthori
+        private void SetupFileServer(IApplicationBuilder builder, IFileProvider files)
         {
-            branch.Use(async (context, next) =>
-            {
-                if (!context.User.Identity.IsAuthenticated)
-                {
-                    await context.ChallengeAsync(new AuthenticationProperties()
-                    {
-                        // https://github.com/aspnet/Security/issues/1730
-                        // Return here after authenticating
-                        RedirectUri = context.Request.PathBase + context.Request.Path + context.Request.QueryString
-                    });
-                    return;
-                }
-
-                await next();
-            });
-            branch.UseFileServer(new FileServerOptions()
+            builder.UseFileServer(new FileServerOptions()
             {
                 EnableDirectoryBrowsing = true,
                 FileProvider = files
             });
         }
 
-        // Policy based authorization, requests must meet the policy criteria to be get access to the resources.
-        private static void MapImperativeFiles(IAuthorizationService authorizationService, IApplicationBuilder branch, PhysicalFileProvider files)
+        private static void SetFileEndpoint(HttpContext context, PhysicalFileProvider files, PathString filePath, string policy)
         {
-            branch.Use(async (context, next) =>
+            var fileSystemInfo = GetFileSystemInfo(files, filePath);
+            if (fileSystemInfo != null)
             {
-                var fileInfo = files.GetFileInfo(context.Request.Path);
-                AuthorizationResult result = null;
-                if (fileInfo.Exists)
-                {
-                    result = await authorizationService.AuthorizeAsync(context.User, fileInfo, "files");
-                }
-                else
-                {
-                    // https://github.com/aspnet/Home/issues/2537
-                    var dir = files.GetDirectoryContents(context.Request.Path);
-                    if (dir.Exists)
-                    {
-                        result = await authorizationService.AuthorizeAsync(context.User, dir, "files");
-                    }
-                    else
-                    {
-                        context.Response.StatusCode = StatusCodes.Status404NotFound;
-                        return;
-                    }
-                }
+                var metadata = new List<object>();
+                metadata.Add(fileSystemInfo);
+                metadata.Add(new AuthorizeAttribute(policy));
+
+                var endpoint = new Endpoint(
+                    c => Task.CompletedTask,
+                    new EndpointMetadataCollection(metadata),
+                    context.Request.Path);
+
+                context.Features.Set<IEndpointFeature>(new EndpointFeature(endpoint));
```

Comment:
This warrants a SetEndpoint extension method that creates the dummy feature for you internally. We do this for other features already. And while you're at it add the matching GetEndpoint extension.

Reply:
Good suggestion! How would `SetEndpoint` work if there is already an `IEndpointFeature` set? Would it update the Endpoint property and leave the same feature, or would it always set a new feature with endpoint.

I ask because the implementation that routing sets implements multiple feature interfaces, e.g. `IEndpointFeature` and `IRouteValuesFeature` and `IRouterFeature` (for legacy).

Reply:
I would not expect it to replace the feature implementation, only the endpoint.